sklearn/utils/validation.py

```diff
@@ -780,9 +780,15 @@ def check_array(
     elif hasattr(array, "iloc") and hasattr(array, "dtype"):
         # array is a pandas series
         pandas_requires_conversion = _pandas_dtype_needs_early_conversion(array.dtype)
-        if pandas_requires_conversion:
-            # Set to None, to convert to a np.dtype that works with array.dtype
+        from pandas.api.types import is_extension_array_dtype, is_categorical_dtype
+
+        if (
+            pandas_requires_conversion and is_extension_array_dtype(array)
+        ) or is_categorical_dtype(array):
+            # Set to None, to let asarray() work out the best dtype
             dtype_orig = None
+        else:
+            dtype_orig = np.result_type(array.dtype)
```

Comment:
This is happening because pandas is casting the bool series to a float ndarray:

```python
import pandas as pd
import numpy as np

a = pd.Series([True, False])

a.astype(None)
# 0    1.0
# 1    0.0
# dtype: float64

# Interesting this returns a bool
np.asarray(a).dtype
# dtype('bool')
```

For this PR, I think we just need to special case bool so we force `astype` to return a bool:

```python
# array is a pandas series
pandas_requires_conversion = _pandas_dtype_needs_early_conversion(array.dtype)
if pandas_requires_conversion:
    if array.dtype == bool:
        # Panda's astype configures a bool Series to a float one, here
        # we force it to be a bool
        dtype_orig = array.
    else:
        # Set to None to let array.astype work out the best dtype
        dtype_orig = None
```

Reply:
Oh this is already out dated.